File: src/main/study/spark/scalaVersion/sql/RDD2DataFrameReflection.scala
package spark.scalaVersion.sql

import org.apache.spark.{SparkConf, SparkContext, sql}
//基于反射的RDD到DF需要用extends app 不能用def main方法
object RDD2DataFrameReflection extends App {

    val conf = new SparkConf()
      .setAppName("WordCount")
      .setMaster("local")
    val sc = new SparkContext(conf)
    val sqlContext  = new sql.SQLContext(sc)

    import sqlContext.implicits._

    case class Student(id:Int,name:String,age:Int)
    val students =  sc.textFile("/Users/xy/Desktop/spark_study/src/main/resources/students.txt")
      .map(line => line.split(",")).map(arr => Student(arr(0).trim().toInt,arr(1),arr(2).trim().toInt))




    val studentDF = students.toDF()

    studentDF.registerTempTable("students")

    val teenagerDF = sqlContext.sql("select * from students where age <= 18")

    val teenagerRDD = teenagerDF.rdd

//  按照顺序排列，java顺序改变了
    teenagerRDD.map(row => Student(row(0).toString.toInt,row(1).toString,row(2).toString.toInt)).collect()
      .foreach(stu => println(stu.id,stu.name,stu.age))

//row getAs得到指定的列
    teenagerRDD.map(row => Student(row.getAs[Int]("id"),row.getAs[String]("name"),row.getAs[Int]("age"))).collect()
      .foreach(stu => println(stu.id,stu.name,stu.age))

//  row getValueMap 得到指定的几列，返回是个map
    teenagerRDD.map(row =>
      {
        val map =  row.getValuesMap[Any](Array("id","name","age"));
        Student(map("id").toString.toInt,map("name").toString,map("age").toString.toInt)

      }).collect().foreach(stu => println(stu.id,stu.name,stu.age))
  
}

File: src/main/study/spark/javaVersion/sql/RDD2DataFrameReflection.java
package spark.javaVersion.sql;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.Function;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SQLContext;


import java.util.List;

public class RDD2DataFrameReflection {
    public static void RDD2DataFrameReflection(){
        SparkConf conf = new SparkConf()
                .setAppName("RDD2DataFrameReflection")
                .setMaster("local");
        JavaSparkContext sc = new JavaSparkContext(conf);
        SQLContext sqlContext = new SQLContext(sc);
        JavaRDD<String> lines = sc.textFile("/Users/xy/Desktop/spark_study/src/main/resources/students.txt");
        JavaRDD<Student> students = lines.map(new Function<String, Student>() {
            @Override
            public Student call(String line) throws Exception {
                String[] lineSplited = line.split(",");
                Student stu = new Student();
                stu.setId(Integer.valueOf(lineSplited[0].trim()));
                stu.setName(lineSplited[1]);
                stu.setAge(Integer.valueOf(lineSplited[2].trim()));
                return stu;
            }
        });
//        使用反射方式，将RDD转化为DataFrame（将class传入）
        Dataset studentDF = sqlContext.createDataFrame(students,Student.class);
//        注册临时表
        studentDF.registerTempTable("students");
        Dataset teenagerDF = sqlContext.sql("select * from students where age <= 18");
//        将查询出来的DF转化为RDD
        JavaRDD<Row>  teenagerRDD = teenagerDF.javaRDD();
        JavaRDD<Student> teenagerStudentRDD = teenagerRDD.map(new Function<Row, Student>() {
            @Override
            public Student call(Row row) throws Exception {
                Student stu = new Student();
                stu.setAge(row.getInt(0));
                stu.setId(row.getInt(1));
                stu.setName(row.getString(2));
                return stu;
            }
        });

        List<Student> studentList = teenagerStudentRDD.collect();

        for (Student stu:studentList){
            System.out.println(stu);
        }
    }
}
